src/oop/programs_with_interesting_classes/medium/q1/MainTest.java

```diff
@@ -0,0 +1,505 @@
+                String.format(incorrectReturnType, getPartsMethodName));
+
+        /* Create instance to call method on */
+        Object[][] constructorArguments = {
+                {"Katie", String.class},
+                {34.2223, double.class}
+        };
+        Object carInstance = car.createInstance(constructorArguments);
+
+        /* Set parts field to passed value */
+        String[] partsCopy = new String[parts.length];
+        System.arraycopy(parts, 0, partsCopy, 0, parts.length);
+        car.setFieldValue(carInstance, partsCopy, stringArrayAttributeName);
+
+        /* Call getParts on car */
+        Object getNameOutput = car.callMethod(getPartsMethodName, carInstance);
+
+        /* Check result */
+        String incorrectUpdate = "Your %s method does not return the value of the %s attribute.";
+        _assertArrayEquals(parts, getNameOutput,
+                String.format(incorrectUpdate, getPartsMethodName, stringArrayAttributeName));
+    }
+
+    private static Stream<Arguments> toStringInputProvider() {
+        return Stream.of(
+                Arguments.of("Another Random Car", 0.51, new String[]{"Wheel", "Motor", "Das Auto"}),
+                Arguments.of("Volkswagen Golf", 345.12, new String[]{""}),
+                Arguments.of("Fiat 500", 0.01, new String[]{}),
+                Arguments.of("Dodge Grand Caravan", 187.66,
+                        new String[]{"Wheels", "Chassis", "Steering", "BodyKit", "Windows", "Doors", "Seats"})
+        );
+    }
+
+    private String expectedToString(String name, double speed, String[] parts) {
+        return String.format("Car{name='%s', speed=%.2f, parts=%s}", name, speed, Arrays.toString(parts));
+    }
+
+    @ParameterizedTest
+    @MethodSource("toStringInputProvider")
+    public void correctToStringMethod(String name, double speed, String[] parts) throws Throwable {
+        /* Check method definition */
+        String incorrectDefinition = "Your %s class is missing the %s method. Make sure it is defined, spelt correctly, and has the correct parameters.";
+        String incorrectModifier = "The %s method does not have the correct visibility modifier.";
+        String incorrectReturnType = "The %s method does not have the correct return type.";
+        assertTrue(car.hasMethod("toString", null),
+                String.format(incorrectDefinition, carClassName, "toString"));
+        assertTrue(car.hasModifier("toString", null, "public"),
+                String.format(incorrectModifier, "toString"));
+        assertTrue(car.hasReturnType("toString", null, String.class),
+                String.format(incorrectReturnType, "toString"));
+
+        /* Initialize object for toString test */
+        Object[][] constructorArguments = {
+                {"Kia Soul", String.class},
+                {11.43, double.class}
+        };
+        Object carInstance = car.createInstance(constructorArguments);
+
```

Comment:
I think here you are mixing the worksTogether and toString method I think its better to test them separately.

Reply:
I am not sure what you mean. I have manually set the fields using the `setFieldValue` method in `ObjectTest`